node/podcontroller.go

```diff
@@ -155,6 +157,8 @@ type PodControllerConfig struct {
 	ConfigMapInformer corev1informers.ConfigMapInformer
 	SecretInformer    corev1informers.SecretInformer
 	ServiceInformer   corev1informers.ServiceInformer
+
+	PodDeletionPolicy PodDeletionPolicy
```

Comment:
I'm not sure how this is supposed to be used, because for each policy we end up having to check values anyway and we aren't using the interface. This seems to be more trying to make an enum in go using an interface.
If we need an enum, which I think we do, I think it's best use the standard way of doing this in go of using simple values.

Reply:
If we have an enum, then we have two values. The enum + the delaying func. This prevents someone from specifying a delaying func, but it not working because they didn't set the enum correctly.

Reply:
I think that's fine, also not everything requires. The enum tells us the strategy so we know how to proceed, and in some cases of the enum that means we need to look at another field. Even then I don't think we always need to do that.
For common cases we can have an enum value for that. The extra field would be for injecting custom behavior.

Reply:
One last thing. RE: Enum, what if you want to control the delete policy independently from the delay policy. For example, if you want to delay *some* pods, and delete *other* pods. Does it make sense to take the above interface and expose it, rather than having an enum value, and two knobs for controlling delete and delay?

Or is this just too wacky of a use case.

Reply:
If we don't want to make it so that every user needs to implement both functions, then we can add some helper functions that make it so they need to implement one or the other. Since the deletion policy includes both of these, making them fully customizable seems like a "better" option. Otherwise, I think we need to expose these directly.

Reply:
My issue here is it's fundamentally not an interface.

Reply:
Why not always have it be a function, and we can define our own default function which noops.

Reply:
How do you handle a case like "never delete"?
I was thinking we'd end up with a long list of return values to handle different cases.

Reply:
If the function's signature is:
```go
deletionPolicyCallback(*v1.Pod) (shouldDelete bool, delayDelete time.Duration)
```

If you want to never delete, you just return `false` for the first value. This would get checked before processing it in the deletionQ. If you return `true`, and a non-zero deletion value, that delays deletion. If you return `true`, and a zero, or negative duration, that triggers immediate deletion.

Reply:
My worry is this can't possibly handle all cases just the common ones and I can see needing to scrap this in the future.

We should also avoid requiring changes in the core in order to support an edge case.